axum/src/json.rs

```diff
@@ -144,29 +144,42 @@ where
     }
 }
 
-fn json_content_type(headers: &HeaderMap) -> bool {
-    let content_type = if let Some(content_type) = headers.get(header::CONTENT_TYPE) {
-        content_type
-    } else {
-        return false;
-    };
-
-    let content_type = if let Ok(content_type) = content_type.to_str() {
-        content_type
-    } else {
-        return false;
-    };
-
-    let mime = if let Ok(mime) = content_type.parse::<mime::Mime>() {
-        mime
-    } else {
-        return false;
-    };
-
-    let is_json_content_type = mime.type_() == "application"
-        && (mime.subtype() == "json" || mime.suffix().map_or(false, |name| name == "json"));
-
-    is_json_content_type
+fn json_content_type(headers: &HeaderMap) -> Option<mime::Mime> {
+    headers
+    .get(header::CONTENT_TYPE)
+    .map(|x| {
+        x.to_str().unwrap()
+    })
+    .and_then(|x| {
+        x.parse::<mime::Mime>().ok()
+    })
+    .and_then(|m| {
+        let suffix = m
+        .suffix()
+        .map_or(false, |name| {
+            name.eq(&mime::JSON)
+        })
+        .then_some(mime::JSON);
+
+        let type_ = m
+        .type_()
+        .eq(&mime::APPLICATION)
+        .then_some(mime::APPLICATION_JSON);
+
+        let subtype = m
+        .subtype()
+        .eq(&mime::JSON)
+        .then_some(mime::JSON);
+
+        [
+            subtype,
+            suffix,
+        ]
+        .iter()
+        .any(|x| x.is_some())
+        .then_some(type_)
+        .flatten()
+    })
```

Comment:
> In the end it is about twice as fast,

Why?

I think readability is also worse now since it doesn't use `?`. All the combinator functions are hard to follow. Also not sure we need to return an `Option<_>` since we're not using it.

Reply:
Hmm, ok, valid, but, "?" was not used before so I dont know what that means (the return value was a boolean).

Yeah, about readability, its a matter of taste I guess, I like the more functional approach.

You quoted the "twice as fast",
well I ran the `Instant::now`,  with about 20 random headers and its about 50% faster. (dunno if that is a good measurement).

Reply:
> Hmm, ok, valid, but, "?" was not used before so I dont know what that means (the return value was a boolean).

I know we didn't use `?` before, but the way you wrote it is less readable IMO without `?`.

> You quoted the "twice as fast", well I ran the Instant::now, with about 20 random headers and its about 50% faster. (dunno if that is a good measurement).

I'm curious if you did any digging into why its faster. That isn't clear to me and I'm curious to know.

Reply:
> I'm curious if you did any digging into why its faster. That isn't clear to me and I'm curious to know.

No clue, isn't this like compiler things. Maybe its just something local (debug build) and in production its optimized. Or maybe its my naive perf test that isn't done right.

---

I guess one option of this change is that the `fn from_request` could go without the branching using `?` with something like,

`json_content_type(req.headers()).ok_or(MissingJsonContentType)?;`

Reply:
I checked the asm output, The asm code generated by this pr is much less than the original, is really a strange problem.